src/bundle/Core/Command/ExpireUserPasswordsCommand.php

```diff
@@ -0,0 +1,383 @@
+    {
+        $beforeRunningHints = self::BEFORE_RUNNING_HINTS;
+        $this
+            ->setDescription('Expire passwords for selected users.')
+            ->addOption(
+                'user-id',
+                null,
+                InputOption::VALUE_OPTIONAL | InputOption::VALUE_IS_ARRAY,
+                'Expire password for specific User identified by ID'
+            )
+            ->addOption(
+                'user-group-id',
+                null,
+                InputOption::VALUE_OPTIONAL | InputOption::VALUE_IS_ARRAY,
+                'Expire passwords of all users assigned to specific User Group'
+            )
+            ->addOption(
+                'user-content-type-identifier',
+                null,
+                InputOption::VALUE_OPTIONAL | InputOption::VALUE_IS_ARRAY,
+                'Expire passwords of all users based on specific Content Type'
+            )
+            ->addOption(
+                'force',
+                null,
+                InputOption::VALUE_NONE,
+                'Perform setting user passwords as expired'
+            )
+            ->addOption(
+                'batch-size',
+                null,
+                InputOption::VALUE_OPTIONAL,
+                'Number of users to process at once',
+                self::DEFAULT_BATCH_SIZE
+            )
+            ->addOption(
+                'password-ttl',
+                null,
+                InputOption::VALUE_OPTIONAL,
+                'After how many days passwords expire. Set when Content Type needs to be updated.',
+                self::DEFAULT_PASSWORD_TTL
+            )
+            ->setHelp(
+                <<<EOT
+The command <info>%command.name%</info> expires passwords of specific users. 
+Use this tool wisely as next time affect users tries to log in, they will be forced to set a new password.
+Note: This script can potentially run for a very long time, and in Symfony dev environment it will consume memory exponentially with size of dataset.
+
+{$beforeRunningHints}
+EOT
+            );
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $batchSize = $input->getOption('batch-size');
+        $userIds = $input->getOption('user-id');
+        $userGroupIds = $input->getOption('user-group-id');
+        $force = $input->getOption('force');
+        $passwordTTL = $input->getOption('password-ttl');
+
+        if (!empty($userIds) && !empty($userGroupIds)) {
+            throw new InvalidArgumentException('You cannot use --user-id and --user-group-id options at once.');
+        }
+
+        $userContentTypeIdentifiers = $input->getOption('user-content-type-identifier');
+
+        $this->summarizeSearchCriteria($output, $userIds, $userGroupIds, $userContentTypeIdentifiers);
+
+        $criteria = $this->getCriteria($userIds, $userGroupIds, $userContentTypeIdentifiers);
+
+        $filter = new Filter(new LogicalAnd($criteria));
```

Comment:
Filter is fluent so you don't have to wrap it artificially.
This and `getCriteria` above could be:
```php
        $filter = new Filter();
        if (!empty($userIds)) {
            $filter->andWithCriterion(new Query\Criterion\ContentId($userIds));
        }

        if (!empty($userGroupIds)) {
            $filter->andWithCriterion(new Query\Criterion\ParentLocationId($userGroupIds));
        }

        if (!empty($userContentTypeIdentifiers)) {
            $filter->andWithCriterion(new Query\Criterion\ContentTypeIdentifier($userContentTypeIdentifiers));
        }
```

It's fine to group this into a method, like you did (due to the mentioned cognitive complexity), but then how about
```php
        $filter = new Filter();
        $this->applyCriteria($filter, $userIds, $userGroupIds, $userContentTypeIdentifiers);
```
?